misc/package/Data Files/MWSE/core/mcm/mcm.lua

```diff
@@ -133,61 +127,65 @@ end
+-- This will be done via the `__index` metamethod as follows:
+--	1. The first time a `create` function is called, it will trigger the `mcm.getComponent` function.
+--		- This function will search the `mcm\\components` folder until it finds the relevant component.
+--		- It will then cache that component so it only needs to search the filetree once.
+--		- The component will be returned by `getComponent` (if it exists)
+--	2. If no component is returned by `mcm.getComponent`, a similar search will be done on `mwseMCMVariable`s.
+--  3. If a `mwseMCMComponent` or a `mwseMCMVariable` is returned by Step 1 or Step 2., we will:
+--		- Create a new function processes the given data and returns a new instance of the relevant class.
+--		- Store this new function in the `mcm` table (so this whole process only happens one time per call to `create<Component|Variable>`).
+--		- Return this new function.
+setmetatable(mcm, {__index=function (_, key) ---@param key string
+	if key:sub(1, strLengthCreate) ~= "create" then return end
+
+	local className = key:sub(strLengthCreate + 1)
+	
+	-- First check if it's a component.
+	local componentClass = fileUtils.getComponentClass(className)
+	if componentClass then
+
+		-- Store the function so we don't have to recreate it every time.
+		mcm[key] = function(param1, param2)
+			local data, parent = param1, nil
+			if param2 then
+				data = param2
+				-- Add check for mcm field to deal with using `:` instead of `.`
+				if param1 ~= mcm then
+					parent = param1
+				end
 			end
-
-			if component then
-				--- @cast component mwseMCMComponent
-				return function(param1, param2)
-					local parent = nil
-					local data = nil
-					if param2 then
-						parent = param1
-						data = param2
-					else
-						data = param1
-					end
-					if not data then
-						data = "---"
-					end
-					if type(data) == "string" then
-						if component.componentType == "Template" then
-							data = { name = data }
-						else
-							data = { label = data }
-						end
-					end
-					data.class = class
-
-					component = component:new(data)
-					-- Add check for mcm field to deal with using `:` instead of `.`
-					if parent and parent.noParent ~= true then
-						component:create(parent)
-					end
-					return component
+			data = data or "---"
+			if type(data) == "string" then
+				if componentClass.componentType == "Template" then
+					data = { name = data }
+				else
+					data = { label = data }
 				end
 			end
+			local component = componentClass:new(data)
+			if parent then
+				component:create(parent)
+			end
+			return component
 		end
+		return mcm[key]
 	end
+	-- Now check if it's a variable.
+	local variableClass = fileUtils.getVariableClass(className)
+	if variableClass then
+		-- Store the function so we don't have to recreate it every time.
+		mcm[key] = function(param1, param2)
+			return variableClass:new(param2 or param1)
+		end
+		return mcm[key]
+	end
```

Comment:
Again, you can reduce the nesting level by inverting the if condition.

Reply:
In this particular case I ended up not inverting the if condition (in the original PR) because I thought it made the `variableClass` part of the code more symmetric with the `componentClass` part of the function. i.e., the function has two main chunks: `if componentClass then ... end` and then later `if variableClass then ... end`, and the symmetrical `if <class> then ... end` checks helped me mentally break up the function into two parts.

I'll make the change you suggested. I just wanted to mention this so that I could ask about the more general setting so that I know for next time: is it generally better to prioritize reducing the nesting level vs maintaining "symmetry" with prior code blocks?

Reply:
I'd say your reasoning is also valid. I may have overdone it with the nesting reduction 😀.